File: NC/products-of-array-discluding-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        totalProduct = 1
        hasZero = 0

        for i in nums:
            if i != 0:
                totalProduct *= i
            else:
                hasZero += 1

        ans = []

        for i in nums:
            if hasZero == 0:
                ans.append(int(totalProduct/i))
            else:
                if hasZero == 1:
                    if i == 0:
                        ans.append(int(totalProduct))
                    else:
                        ans.append(0)
                else:
                    ans.append(0)

        return ans

File: NC/java/ProductsOfArrayDiscludingSelf.java
package NC.java;

public class ProductsOfArrayDiscludingSelf {
    public int[] productExceptSelf(int[] nums) {
        int zeroCount = 0;
        int totalProduct = 1;

        for (var num : nums) {
            if (num == 0) {
                zeroCount++;
            } else {
                totalProduct *= num;
            }
        }

        int[] ans = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            int curr = nums[i];
            if (zeroCount > 1) {
                ans[i] = 0;
            } else if (zeroCount == 1 && curr != 0) {
                ans[i] = 0;
            } else if (zeroCount == 1 && curr == 0) {
                ans[i] = totalProduct;
            } else {
                ans[i] = totalProduct / curr;
            }
        }

        return ans;
    }
}
